### PR TITLE
[distributed] Fix max_seqlen mismatch in ring attention backward

### LINKED ISSUE #163075: [Bug] max_seqlen parameter can not align with input in ring attention backward
### 🐛 Describe the bug

Currently, the implementation of ring attention in PyTorch will dispatch multiple sub-attention (one per iteration) in the forward operator (https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/experimental/_attention.py#L466)
However the rest list which will be returned only have one (always be the last iteration's rest) https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/experimental/_attention.py#L475
The rest list contains max_seqlen_q and max_seqlen_k which will be used by ring attenion backward. 

The problem is: in backward, the max_seqlen_q and max_seqlen_k will always be the return value of the last sub-attenion which's q or kv input may be chunked, and they can not align with all the sub ops dispatched in backward eg. in the 0 iteration the full q kv will be caculated and the max_seqlen_q may be less than the input's seqlen (https://github.com/pytorch/pytorch/blob/release/2.8/torch/distributed/tensor/experimental/_attention.py#L709)

### Versions

main

cc @ezyang @gchanan @kadeng @msaroufim @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @dcci @aditvenk @weifengpy @tianyu-l @XilunWu @SherlockNoMad @ppwwyyxx @drisspg @liangel-02 @howardzhang-cv @H-Huang

### PR DESCRIPTION
**Pull request by RohitRathore1**
With load balancing enabled, the forward loop overwrites `rest` (which contains max_q/max_k) on every iteration, returning values from the last iteration that may have used chunked q/k. The backward then passes these fixed values to every iteration even though each iteration uses differently-shaped tensors.

Save `rest` from iteration 0 (which always uses full q/k/v) and update max_q/max_k per backward iteration to match the actual tensor shapes.

Fixes #163075

cc @ezyang @gchanan @kadeng @msaroufim

### CODE REVIEW

**Review by drisspg — APPROVED** [review]
This looks right to me, however only if we are not using dropout or varlen which from a quick pass  I dont think we are. Can we assert that somewhere and raise a value error

**Comment by drisspg** [comment]
@claude Please do a deep review focused on subtle correctness bugs and regressions that are easy for human reviewers to miss. Prioritize findings over summary. I dont care about anything that isnt a real bug drop all the noise but that, if you there isnt anything to report just say so with a small comment

**Review by fegin — APPROVED** [review]

**Comment by RohitRathore1** [comment]
hi @fegin and @drisspg these tests failures are not related to my changes

**Comment by RohitRathore1** [comment]
## CI Failure Analysis

All 3 failing jobs are unrelated to this PR (which only modifies `torch/distributed/tensor/experimental/_context_parallel/_attention.py`):

**1. `dtensor-test / test-osdc` — `test_compile_bernoulli` timeout**
- The failing test is `DistTensorRandomOpCompileTest::test_compile_bernoulli` in `test_random_ops.py` — a random ops compilation test that timed out after 300s in CI.
- Passes locally in ~23s. This is a CI flake, not related to attention code.

**2. `torchtitan-x-pytorch-test / test-osdc` — torchtitan integration**
- Torchtitan feature integration test infra issue. No test assertion failure in logs — the "Test" step is marked failed but logs show training steps running normally.
- Unrelated to distributed attention changes.

**3. `macos-py3-arm64 / test (mps)` — macOS MPS test**
- macOS MPS backend test. This PR only touches CUDA distributed attention code. Completely unrelated.

**Local verification:**
```
$ python test/distributed/tensor/test_random_ops.py -v -k "test_compile_bernoulli"
Ran 2 tests in 23.466s — OK

$ python test/distributed/tensor/test_attention.py -v -k "test_ring_attention_sdpa"
Ran 2 tests in 19.852s — OK
```

cc: @fegin

**Comment by fegin** [comment]
There is a known issue of torchtitan-test and is being fixed atm. Let's wait for the fix and rebase again.

**Comment by fegin** [comment]
After https://github.com/pytorch/pytorch/pull/187067 is landed, we can rebase to verify if CIs are green.
